src/Confluent.SchemaRegistry/SchemaId.cs

```diff
@@ -0,0 +1,298 @@
+// Copyright 2025 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using Confluent.Kafka;
+
+namespace Confluent.SchemaRegistry;
+
+/// <summary>
+///     Represents a schema ID or GUID.
+/// </summary>
+public class SchemaId
+{
+    public const string KEY_SCHEMA_ID_HEADER = "__key_schema_id";
+    public const string VALUE_SCHEMA_ID_HEADER = "__value_schema_id";
+    public const byte MAGIC_BYTE_V0 = 0;
+    public const byte MAGIC_BYTE_V1 = 1;
+
+    private const int DefaultInitialBufferSize = 1024;
+
+    private static List<int> _defaultIndex = new List<int> { 0 };
+
+    /// <summary>
+    ///     The schema type.
+    /// </summary>
+    public SchemaType SchemaType { get; }
+
+    /// <summary>
+    ///     The schema ID.
+    /// </summary>
+    public int? Id { get; set;  }
+
+    /// <summary>
+    ///     The schema GUID.
+    /// </summary>
+    public Guid? Guid { get; set;  }
+
+    /// <summary>
+    ///     The schema GUID.
+    /// </summary>
+    public List<int> MessageIndexes { get; set;  }
+
+    public SchemaId(SchemaType schemaType)
+    {
+        SchemaType = schemaType;
+    }
+
+    public SchemaId(SchemaType schemaType, int id, Guid guid)
+    {
+        SchemaType = schemaType;
+        Id = id;
+        Guid = guid;
+    }
+
+    public SchemaId(SchemaType schemaType, int id, string guid)
+    {
+        SchemaType = schemaType;
+        Id = id;
+        Guid = guid != null ? System.Guid.Parse(guid) : null;
+    }
+
+    public Stream FromBytes(Stream stream)
+    {
+        var reader = new BinaryReader(stream);
+        var magicByte = reader.ReadByte();
+        if (magicByte == MAGIC_BYTE_V0)
+        {
+            Id = IPAddress.NetworkToHostOrder(reader.ReadInt32());
+        }
+        else if (magicByte == MAGIC_BYTE_V1)
+        {
+            Guid = Utils.GuidFromBigEndian(reader.ReadBytes(16));
+        }
+        else
+        {
+            throw new InvalidDataException($"Invalid magic byte: {magicByte}");
+        }
+
+        if (SchemaType == SchemaType.Protobuf)
+        {
+            var size = stream.ReadVarint();
+            if (size == 0)
+            {
+                MessageIndexes = _defaultIndex;
+            }
+            else
+            {
+                MessageIndexes = new List<int>();
```

Comment:
We're not actually using this for anything in the deserialization flow; nevertheless, this array is allocated for every message we consume. I'm wondering if it would be possible to detach MessageIndexes from the type representing the schema ID, since MessageIndexes are specific to Protobuf serialization anyway.

Reply:
This is only instantiated for Protobuf messages that don't used the first message type in the schema, which is uncommon